### PR TITLE
for not-grpc request, should not get grpc-status -- fix issue #27201

### PR DESCRIPTION
**Pull request by johnzheng1975**
for not-grpc request, should not get grpc-status -- fix issue #27201

Commit Message:  Signed-off-by: John Zheng  <john.zheng@hp.com>
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by johnzheng1975** [comment]
Not sure why test is fail, Thanks a lot if someone can help to change directly. Thanks.

**Comment by zirain** [comment]
> Not sure why test is fail, Thanks a lot if someone can help to change directly. Thanks.
```console
test/common/formatter/substitution_formatter_test.cc:2872: Failure
Value of: formatter.formatValue(request_header, response_header, response_trailer, stream_info, body, AccessLog::AccessLogType::NotSet)
Expected: proto eq 32-byte object <90-17 54-01 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 40-DE E4-40 00-00 00-00 02-00 00-00>
  Actual: 32-byte object <90-17 54-01 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 40-DE E4-40 00-00 00-00 01-00 00-00> (of type google::protobuf::Value), 
==========================Expected proto:===========================
number_value: 42738
------------------is not equal to actual proto:---------------------
null_value: NULL_VALUE
```

**Comment by johnzheng1975** [comment]
> > Not sure why test is fail, Thanks a lot if someone can help to change directly. Thanks.
> 
> ```
> test/common/formatter/substitution_formatter_test.cc:2872: Failure
> Value of: formatter.formatValue(request_header, response_header, response_trailer, stream_info, body, AccessLog::AccessLogType::NotSet)
> Expected: proto eq 32-byte object <90-17 54-01 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 40-DE E4-40 00-00 00-00 02-00 00-00>
>   Actual: 32-byte object <90-17 54-01 00-00 00-00 00-00 00-00 00-00 00-00 00-00 00-00 40-DE E4-40 00-00 00-00 01-00 00-00> (of type google::protobuf::Value), 
> ==========================Expected proto:===========================
> number_value: 42738
> ------------------is not equal to actual proto:---------------------
> null_value: NULL_VALUE
> ```

Happy to see you again in envoy community. Thanks @zirain 
Yes, I saw it. But I do not know how to resolve it.
I added an condition, onlyGrpcStatusFormatter when it isGrpcRequestHeaders.
This makes test fail because there is no grpc content-type. 

In order to resolve this issue, I added: 
Http::TestRequestHeaderMapImpl request_header{{"content-type", "application/grpc+proto"}};
I suppose this can work, but it failed.

File: test/common/formatter/substitution_formatter_test.cc
Comment (ohadvano) [inline test/common/formatter/substitution_formatter_test.cc:2683]: Looking at isGrpcRequestHeaders:
https://github.com/envoyproxy/envoy/blob/5523257c7059f0748611b0d83cfa7c28268d0df9/source/common/grpc/common.cc#L55

You would also need to set a path for the request.

Comment (johnzheng1975) [reply test/common/formatter/substitution_formatter_test.cc:2683]: Thanks a lot! You are right! @ohadvano

**Comment by johnzheng1975** [comment]
Met an format issue, 
##[error]Publishing build artifacts failed with an error: Not found PathtoPublish: /home/vsts/work/1/a/fix_format.diff
Also has error like: https://source.cloud.google.com/results/invocations/ca1689c5-9e4d-43f2-8824-7847845ae87f/log

**Review by ohadvano — COMMENTED** [review]
I think there can also be tests to check the new code path where there's an empty value for the gRPC status in case the request was not gRPC, lgtm otherwise

**Comment by johnzheng1975** [comment]
Thanks for your suggestions, Added tests for not gRPC requests. 
Please help to review again. @ohadvano

**Review by ohadvano — COMMENTED** [review]
I think there's a place for a similar test case in access_log_impl_test, as this is where you expect the actual outcome to be different. However I'm not sure if it's mandatory, I assume a maintainer will soon review this.

**Comment by johnzheng1975** [comment]
Thanks @ohadvano for your time. :-) 
I just think add another test case to access_log_impl_test may overlap.
Waiting for an maintainer to review.

**Comment by wbpcode** [comment]
> I'm wondering why there's a grpc-status response header for non-grpc request to begin with, any chance you managed to debug this?

@johnzheng1975 same quesion with @ohadvano. Why non grpc request's response will have a grpc status header? 
I think in the current implementation, envoy has checked the response header to see if a grpc status header exist and only print it to log when the header exist.

```
  const auto grpc_status =
      Grpc::Common::getGrpcStatus(response_trailers, response_headers, info, true);
  if (!grpc_status.has_value()) {
    return absl::nullopt;
  }


absl::optional<Status::GrpcStatus>
Common::getGrpcStatus(const Http::ResponseHeaderOrTrailerMap& trailers, bool allow_user_defined) {
  const absl::string_view grpc_status_header = trailers.getGrpcStatusValue();
  uint64_t grpc_status_code;

  if (grpc_status_header.empty()) {
    return absl::nullopt;
  }
  if (!absl::SimpleAtoi(grpc_status_header, &grpc_status_code) ||
      (grpc_status_code > Status::WellKnownGrpcStatus::MaximumKnown && !allow_user_defined)) {
    return {Status::WellKnownGrpcStatus::InvalidCode};
  }
  return {static_cast<Status::GrpcStatus>(grpc_status_code)};
}
```

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks.

**Comment by wbpcode** [comment]
Wait the CI to pass and defer to @yanavlasov for final review or merge.

File: source/common/formatter/substitution_formatter.cc
Comment (yanavlasov) [inline source/common/formatter/substitution_formatter.cc:1853]: Thanks you for your change. It looks good. However given that it is a user visible change it requires a runtime feature protection. If somebody has production monitoring that was affected by this change, they could temporarily disable the new behavior until they adjust. Here is a good example of how to add this: https://github.com/envoyproxy/envoy/pull/26764

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/common/formatter/substitution_formatter.cc
Comment (johnzheng1975) [reply source/common/formatter/substitution_formatter.cc:1853]: @yanavlasov , Thanks for your suggestion, it make sense. 👍 
I like this protection, will add this change.

**Comment by johnzheng1975** [comment]
Hmm, I miss one sign-off.
it said:
```
Summary
Commit sha: [18ad8c2](https://github.com/envoyproxy/envoy/pull/27219/commits/18ad8c2cf3317eb426b96ee33e3334ac2fe2d7f5), Author: Zheng, Committer: Zheng; The sign-off is missing.
```
Can anyone suggest how to resolve it? or I create a new PR? Thanks a lot.

**Comment by johnzheng1975** [comment]
Will close it sooner.
Clean the code in new PR: https://github.com/envoyproxy/envoy/pull/27387

**Pull request closed without merging**
